README.md
# volume_control

A flutter plugin to programmatically adjust the device's volume on Android and iOS.

## API Example
``` dart
// Import package
import 'package:volume_control/volume_control.dart';

// Get the current volume, min=0, max=1
double _val = await VolumeControl.volume;

// Set the new volume value, between 0-1
VolumeControl.setVolume(0.4);
```

## Example
``` dart
import 'package:flutter/material.dart';
import 'dart:async';
import 'package:volume_control/volume_control.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {

  @override
  void initState() {
    super.initState();
    initVolumeState();
  }

  //init volume_control plugin
  Future<void> initVolumeState() async {
    if (!mounted) return;

    //read the current volume
    _val = await VolumeControl.volume;
    setState(() {
    });
  }

  double _val = 0.5;
  Timer timer;
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Center(
          child: Slider(value:_val,min:0,max:1,divisions: 100,onChanged:(val){
            _val = val;
            setState(() {});
            if (timer!=null){
              timer.cancel();
            }
            
            //use timer for the smoother sliding 
            timer = Timer(Duration(milliseconds: 200), (){VolumeControl.setVolume(val);});
            
            print("val:${val}");
          })
        )

      ),
    );
  }
}
```

CHANGELOG.md
## 0.1.5 - May 21rd, 2021.
* fixed bugs


## 0.1.4 - May 21rd, 2021.
* fixed null safety

## 0.1.3 - May 23rd, 2020.
* fixed code format

## 0.1.2 - May 23rd, 2020.
* fixed code format

## 0.1.1 - May 23rd, 2020.
* fixed on comments

## 0.1.0 - May 23rd, 2020.
* fixed some errors and add some comments

## 0.0.1 - May 22rd, 2020.
* initial release.


ios/Classes/VolumeControlPlugin.m
#import "VolumeControlPlugin.h"
#import "MediaPlayer/MPVolumeView.h"
//AVFoundation.AVFAudio.AVAudioSession
#import <AVFoundation/AVFoundation.h>

@implementation VolumeControlPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"volume_control"
            binaryMessenger:[registrar messenger]];
  VolumeControlPlugin* instance = [[VolumeControlPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}


- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  NSLog(@"method call");

  if ([@"dispose" isEqualToString:call.method]) {
      NSLog(@"VolumeControlPlugin dispose");

  }else if ([@"setVolume" isEqualToString:call.method]) {

      @autoreleasepool {
          @try {
              NSNumber *vol = call.arguments[@"vol"];
              float _vol = vol.floatValue;
              if (_vol>1)_vol=1;
              if (_vol<0)_vol=0;

              MPVolumeView *volumeView = [[MPVolumeView alloc] init];
              UISlider *volumeViewSlider = nil;
              
              for (UIView *view in volumeView.subviews) {
                  if ([view isKindOfClass:[UISlider class]]) {
                      volumeViewSlider = (UISlider *)view;
                      break;
                  }
              }
              
              dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.0001 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
                  volumeViewSlider.value = _vol;
                  //NSLog(@"setting finished!");
              });

          } @catch (NSException *exception) {
              NSLog(@"%@", exception);
          }
          
      }

      
  }else if ([@"getVolume" isEqualToString:call.method]) {

     @autoreleasepool {
         @try {
             [[AVAudioSession sharedInstance] setActive: YES error: nil];
             float vol = [[AVAudioSession sharedInstance] outputVolume];
             
             dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.0001 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
                 //NSLog(@"get val:%f",volumeViewSlider.value);
                 result([NSNumber numberWithFloat:vol]);
                 
             });

         } @catch (NSException *exception) {
             NSLog(@"%@", exception);
         }
         
     }

  } else {
    result(FlutterMethodNotImplemented);
  }
}


@end


android/src/main/java/com/weyiyong/plugins/volume_control/VolumeControlPlugin.java
package com.weyiyong.plugins.volume_control;

import android.content.Context;
import android.media.AudioManager;

import androidx.annotation.NonNull;

import java.util.Map;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** VolumeControlPlugin */
public class VolumeControlPlugin implements FlutterPlugin, MethodCallHandler {
  AudioManager audioManager;
  static Context context;
  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    final MethodChannel channel = new MethodChannel(flutterPluginBinding.getFlutterEngine().getDartExecutor(), "volume_control");
    context = flutterPluginBinding.getApplicationContext();
    channel.setMethodCallHandler(new VolumeControlPlugin());
  }

  public void init(){
    if (audioManager!=null)return;
    //System.out.println("init...:"+context);
    audioManager = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);

  }
  // This static function is optional and equivalent to onAttachedToEngine. It supports the old
  // pre-Flutter-1.12 Android projects. You are encouraged to continue supporting
  // plugin registration via this function while apps migrate to use the new Android APIs
  // post-flutter-1.12 via https://flutter.dev/go/android-project-migration.
  //
  // It is encouraged to share logic between onAttachedToEngine and registerWith to keep
  // them functionally equivalent. Only one of onAttachedToEngine or registerWith will be called
  // depending on the user's project. onAttachedToEngine or registerWith must both be defined
  // in the same class.
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "volume_control");
    channel.setMethodCallHandler(new VolumeControlPlugin());
    context = registrar.context();
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    if (call.method.equals("dispose")) {
    }else if (call.method.equals("getVolume")) {
      init();
      if (audioManager==null){
        //System.out.println("audioManager is null");
        return;
      }

      int max = audioManager.getStreamMaxVolume(AudioManager.STREAM_MUSIC);
      int volume = audioManager.getStreamVolume(AudioManager.STREAM_MUSIC);
      result.success(((double)volume)/max);

    }else if (call.method.equals("setVolume")) {
      init();
      if (audioManager==null){
        //System.out.println("audioManager is null");
        return;
      }
      double vol = (double)((Map)call.arguments).get("vol");
      //System.out.println("setting:"+vol);

      int max = audioManager.getStreamMaxVolume(AudioManager.STREAM_MUSIC);
      //System.out.println("Volume max:"+max);

      audioManager.setStreamVolume(AudioManager.STREAM_MUSIC, (int)(vol*max),0);
      int volume = audioManager.getStreamVolume(AudioManager.STREAM_MUSIC);
      //System.out.println("current vol:"+volume);

    } else {
      result.notImplemented();
    }
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';
import 'package:volume_control/volume_control.dart';


//start entry
void main() => runApp(MyApp());

//usage example
class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

//usage example
class _MyAppState extends State<MyApp> {

  @override
  //init the plugin
  void initState() {
    super.initState();
    initVolumeState();
  }

  //init volume_control plugin
  Future<void> initVolumeState() async {
    if (!mounted) return;

    //read the current volume
    _val = await VolumeControl.volume;
    setState(() {
    });
  }

  double _val = 0.5;
  Timer? timer;
  @override
  //build a slider for example
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Center(
          child: Slider(value:_val,min:0,max:1,divisions: 100,onChanged:(val){
            _val = val;
            setState(() {});
            if (timer!=null){
              timer?.cancel();
            }
            
            //use timer for the smoother sliding 
            timer = Timer(Duration(milliseconds: 200), (){VolumeControl.setVolume(val);});
            
            print("val:$val");
          })
        )

      ),
    );
  }
}
